add devcontainer

june-fish opened the pull request:
TODO: 

- [x] figure out if we want to set formatting settings
- [x] test on Windows and/or Mac

otherwise tested & working on Ultramarine Linux 42 (Fedora) with podman/podman-docker and MS VS Code

june-fish committed:
add devcontainer

Signed-off-by: june-fish <git@june.fish>

june-fish committed:
add formatting

Signed-off-by: june-fish <git@june.fish>

june-fish commented:
Works on macOS (M1)

remexre commented:
Missing trailing newlines? (Also, the DCO thing seems mad.)

remexre committed:
Adds an mdbook for documentation. (#16)

* Adds an mdbook for documentation.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Wires documentation into the Makefiles.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Makes GitHub Actions build and upload documentation.

Signed-off-by: Nathan Ringo <me@remexre.com>

---------

Signed-off-by: Nathan Ringo <me@remexre.com>
Signed-off-by: june-fish <git@june.fish>

remexre committed:
Adds an mdbook for documentation. (#16)

* Adds an mdbook for documentation.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Wires documentation into the Makefiles.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Makes GitHub Actions build and upload documentation.

Signed-off-by: Nathan Ringo <me@remexre.com>

---------

Signed-off-by: Nathan Ringo <me@remexre.com>
Signed-off-by: june-fish <git@june.fish>

remexre committed:
Removes the on.pull_request on the deploy GitHub action. (#24)

Signed-off-by: Nathan Ringo <me@remexre.com>
Signed-off-by: june-fish <git@june.fish>

remexre commented:
Oh, another checkbox:

- [x] Can LSP be made to work? We might need to use ccls, we use parts of C23 that clangd hasn't implemented yet...

remexre committed:
Early exception handler (#19)

* Changes hart-locals to be stored in SSCRATCH instead of tp.

This lets us use hart-locals inside trap handlers.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Adds documentation page on threads vs hart.

Currently, this only really covers the terminology, but it's our first
real documentation page.

Signed-off-by: Nathan Ringo <me@remexre.com>

* Adds an exception handler that applies early in boot.

Signed-off-by: Nathan Ringo <me@remexre.com>

---------

Signed-off-by: Nathan Ringo <me@remexre.com>
Signed-off-by: june-fish <git@june.fish>

june-fish committed:
newlines lawl
ringo is this a profressional commit message yet

Signed-off-by: june-fish <git@june.fish>

june-fish commented:
Works in Windows (WSL Docker) too

june-fish commented:
> Missing trailing newlines? (Also, the DCO thing seems mad.)

Fixed

june-fish commented:
> Oh, another checkbox:
> 
>     * [ ]  Can LSP be made to work? We might need to use ccls, we use parts of C23 that clangd hasn't implemented yet...

Already using IntelliSense, is this ok or should we use something else?

june-fish committed:
add basic devcontainer doc

Signed-off-by: june-fish <git@june.fish>

remexre commented:
> > Oh, another checkbox:
> > ```
> > * [ ]  Can LSP be made to work? We might need to use ccls, we use parts of C23 that clangd hasn't implemented yet...
> > ```
> 
> Already using IntelliSense, is this ok or should we use something else?

IntelliSense is OK; I was hoping for something GCC-based, but yeah, it doesn't look like anything exists. I did find the Reddit post that misinformed me that ccls could plug into GCC :upside_down_face:

remexre commented on doc/getting-started/devcontainer.md:
```diff
@@ -0,0 +1,7 @@
+
+## Using Visual Studio Code
+
+Install [Visual Studio Code](https://code.visualstudio.com/Download) for your system and install the ["Dev Containers" extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers). After cloning the repo and opening it in VS Code you should be prompted to "Reopen in dev container". If not, you can open the command pallet (Ctrl+Shift+P for Linux/Windows, Cmd+S ...
```
```suggestion
Install [Visual Studio Code](https://code.visualstudio.com/Download) for your system and install the ["Dev Containers" extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers). After cloning the repo and opening it in VS Code you should be prompted to "Reopen in dev container". If not, you can open the command palette (Ctrl+Shift+P for Linux/Windows, Cmd+Shift+P for Mac) and type "Reopen in dev container" which should highlight the correct option (press enter to select it).
```

remexre replied:
Also, one line per sentence leads to much more readable diffs; objections to switching to that?

remexre commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+            coreutils \
+            gcc-riscv-none-elf \
+            git \
+            make \
```
minicom

june-fish committed:
Update doc/getting-started/devcontainer.md

Co-authored-by: remexre <me@remexre.com>
Signed-off-by: june-fish <git@june.fish>

june-fish committed:
add minicom

Signed-off-by: june-fish <git@june.fish>

june-fish committed:
better md formatting

Signed-off-by: june-fish <git@june.fish>

june-fish committed:
Revert "Adds an mdbook for documentation. (#16)"

This reverts commit 98cbf201cd654b7fd6986e74b23d8d7c69ce791e.

Signed-off-by: june-fish <git@june.fish>

june-fish committed:
Merge remote-tracking branch 'origin/trunk' into devcontainer

june-fish committed:
im a responsible fish who can be trusted with git

Signed-off-by: june-fish <git@june.fish>
